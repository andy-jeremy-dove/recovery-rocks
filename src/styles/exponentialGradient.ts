import chroma from 'chroma-js';
import type {ColorValue} from 'react-native';

export type GradientPoint = {
  /**
   * [0; 1]
   */
  offset: number;
  color: ColorValue;
  /**
   * [0; 1]
   */
  opacity: number;
};

/**
 * Compute linear gradient points to emulate an exponential gradient
 * @param fromColor
 * @param fromOpacity `[0; 1]`
 * @param toColor
 * @param toOpacity `[0; 1]`
 * @param ratio `(0; 1)`
 * @param interimPoints
 */
export default function exponentialGradient(
  fromColor: ColorValue,
  fromOpacity: number,
  toColor: ColorValue,
  toOpacity: number,
  ratio = GOLDEN_RATIO_RECIPROCAL,
  interimPoints = 1,
): [GradientPoint, ...GradientPoint[], GradientPoint] {
  const start: GradientPoint = {
    offset: 0,
    color: fromColor,
    opacity: fromOpacity,
  };
  const stop: GradientPoint = {offset: 1, color: toColor, opacity: toOpacity};
  if (
    (fromColor === toColor && fromOpacity === toOpacity) ||
    interimPoints <= 0
  ) {
    return [start, stop];
  }
  const gradient = chroma
    .scale([fromColor as string, toColor as string])
    .cache(false)
    .mode('lrgb')
    .colors(interimPoints + 2, 'hex');
  const internal: GradientPoint[] = [];
  for (let i = 0; i < interimPoints; i++) {
    const point = i + 1;
    const color = gradient[point];
    const opacityRatio = (1 / (interimPoints + 1)) * point;
    const opacity = fromOpacity + (toOpacity - fromOpacity) * opacityRatio;
    const offset = 1 - ratio ** point;
    internal.push({offset, color, opacity});
  }
  return [start, ...internal, stop];
}

export const GOLDEN_RATIO_RECIPROCAL = (Math.sqrt(5) - 1) / 2;
