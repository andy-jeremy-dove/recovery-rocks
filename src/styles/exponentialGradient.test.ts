import {expect, it} from '@jest/globals';

import exponentialGradient from './exponentialGradient';

it('obeys invariants', () => {
  const fromColor = '#0068C8';
  const fromOpacity = 1;
  const toColor = '#263379';
  const toOpacity = 1;
  const interimPoints = 3;
  const gradient = exponentialGradient(
    fromColor,
    fromOpacity,
    toColor,
    toOpacity,
    0.7,
    interimPoints,
  );
  expect(gradient.length).toBe(2 + interimPoints);
  const first = gradient[0];
  const last = gradient[gradient.length - 1];
  expect(first.color).toBe(fromColor);
  expect(first.offset).toBe(0);
  expect(first.opacity).toBe(fromOpacity);
  expect(last.color).toBe(toColor);
  expect(last.offset).toBe(1);
  expect(last.opacity).toBe(toOpacity);
});

it('computes exponential gradient of 3 interim points', () => {
  const gradient = exponentialGradient('#0068C8', 1, '#263379', 0, 0.7, 3);
  expect(gradient).toMatchSnapshot();
});

it('computes exponential gradient of 2 interim points', () => {
  const gradient = exponentialGradient('#0068C8', 1, '#263379', 0, 0.65, 2);
  expect(gradient).toMatchSnapshot();
});
