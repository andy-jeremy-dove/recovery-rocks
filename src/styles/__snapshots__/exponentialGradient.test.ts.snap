// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`computes exponential gradient of 2 interim points 1`] = `
[
  {
    "color": "#0068C8",
    "offset": 0,
    "opacity": 1,
  },
  {
    "color": "#165ab2",
    "offset": 0.35,
    "opacity": 0.6666666666666667,
  },
  {
    "color": "#1f4998",
    "offset": 0.5774999999999999,
    "opacity": 0.33333333333333337,
  },
  {
    "color": "#263379",
    "offset": 1,
    "opacity": 0,
  },
]
`;

exports[`computes exponential gradient of 3 interim points 1`] = `
[
  {
    "color": "#0068C8",
    "offset": 0,
    "opacity": 1,
  },
  {
    "color": "#135eb7",
    "offset": 0.30000000000000004,
    "opacity": 0.75,
  },
  {
    "color": "#1b52a5",
    "offset": 0.51,
    "opacity": 0.5,
  },
  {
    "color": "#214491",
    "offset": 0.657,
    "opacity": 0.25,
  },
  {
    "color": "#263379",
    "offset": 1,
    "opacity": 0,
  },
]
`;
